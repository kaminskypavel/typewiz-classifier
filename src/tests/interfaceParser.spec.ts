import {
	getInterfaceName,
	getInterfaceProperties,
	getPropertyKind,
	getPropertyName,
	getPropertyQuestion,
	parseInterface
} from '../interfaceParser';
import {example1, example2, example3, example4} from './fixtures/mockInterfaces';

describe('#getInterfaceName', () => {
	it('should return the interface name', () => {
		expect(getInterfaceName(example1)).toBe('Example1');
		expect(getInterfaceName(example2)).toBe('Example2');
		expect(getInterfaceName(example3)).toBe('Example3');
		expect(getInterfaceName(example4)).toBe('Example4');
	});
});

describe('#getProperties', () => {
	it('should find the correct number on nodes', () => {
		expect(getInterfaceProperties(example1).length).toEqual(0);
		expect(getInterfaceProperties(example2).length).toEqual(5);
		expect(getInterfaceProperties(example3).length).toEqual(6);
		expect(getInterfaceProperties(example4).length).toEqual(9);
	});
});

describe('#getPropertyName', () => {
	it('should find the correct node names', () => {
		expect(getInterfaceProperties(example1).map((node) => getPropertyName(node))).toEqual([]);

		expect(getInterfaceProperties(example2).map((node) => getPropertyName(node))).toEqual([
			'requestId',
			'orderedDependencies',
			'dependencyUrls',
			'javaScript',
			'paramValues'
		]);

		expect(getInterfaceProperties(example3).map((node) => getPropertyName(node))).toEqual([
			'allowSort',
			'fieldName',
			'fieldType',
			'indexAnalyzer',
			'searchAnalyzer',
			'similarity'
		]);

		expect(getInterfaceProperties(example4).map((node) => getPropertyName(node))).toEqual([
			'minZoom',
			'maxZoom',
			'radius',
			'extent',
			'nodeSize',
			'log',
			'reduce',
			'initial',
			'map'
		]);
	});
});

describe('#getPropertyQuestion', () => {
	it('should find the correct node question mark', () => {
		expect(getInterfaceProperties(example1).map((node) => getPropertyQuestion(node))).toEqual([]);

		expect(getInterfaceProperties(example2).map((node) => getPropertyQuestion(node))).toEqual([
			false,
			true,
			true,
			true,
			false
		]);

		expect(getInterfaceProperties(example3).map((node) => getPropertyQuestion(node))).toEqual([
			true,
			false,
			true,
			true,
			true,
			true
		]);

		expect(getInterfaceProperties(example4).map((node) => getPropertyQuestion(node))).toEqual([
			true,
			true,
			true,
			true,
			true,
			true,
			true,
			true,
			true
		]);
	});
});

describe('#getPropertyKind', () => {
	it('should find the correct node question mark', () => {
		expect(getInterfaceProperties(example1).map((node) => getPropertyKind(node))).toEqual([]);

		expect(getInterfaceProperties(example2).map((node) => getPropertyKind(node))).toEqual([
			'number',
			'string[]',
			'{[id:string]:string}',
			'string',
			'any[]'
		]);

		expect(getInterfaceProperties(example3).map((node) => getPropertyKind(node))).toEqual([
			'boolean',
			'string',
			'Field.FieldTypeEnum',
			'string',
			'string',
			'Field.SimilarityEnum'
		]);

		expect(getInterfaceProperties(example4).map((node) => getPropertyKind(node))).toEqual([
			'number',
			'number',
			'number',
			'number',
			'number',
			'boolean',
			'(accumulated:any,props:any)=>void',
			'()=>any',
			'(props:any)=>any'
		]);
	});
});

describe('#parseInterface', () => {
	it('should parse the interfaces correctly', () => {
		[example1, example2, example3, example4].forEach((example, idx) => {
			expect(parseInterface(example)).toMatchSnapshot(`example${idx + 1}`);
		});
	});
});
