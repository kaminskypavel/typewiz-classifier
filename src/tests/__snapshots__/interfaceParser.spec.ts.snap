// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#parseInterface should parse the interfaces correctly: example1 1`] = `
Object {
  "name": "Example1",
  "structure": Array [],
}
`;

exports[`#parseInterface should parse the interfaces correctly: example2 1`] = `
Object {
  "name": "Example2",
  "structure": Array [
    Object {
      "kind": "number",
      "name": "requestId",
      "question": false,
      "stem": "requestid",
    },
    Object {
      "kind": "string[]",
      "name": "orderedDependencies",
      "question": true,
      "stem": "ordereddepend",
    },
    Object {
      "kind": "{[id:string]:string}",
      "name": "dependencyUrls",
      "question": true,
      "stem": "dependencyurl",
    },
    Object {
      "kind": "string",
      "name": "javaScript",
      "question": true,
      "stem": "javascript",
    },
    Object {
      "kind": "any[]",
      "name": "paramValues",
      "question": false,
      "stem": "paramvalu",
    },
  ],
}
`;

exports[`#parseInterface should parse the interfaces correctly: example3 1`] = `
Object {
  "name": "Example3",
  "structure": Array [
    Object {
      "kind": "boolean",
      "name": "allowSort",
      "question": true,
      "stem": "allowsort",
    },
    Object {
      "kind": "string",
      "name": "fieldName",
      "question": false,
      "stem": "fieldnam",
    },
    Object {
      "kind": "Field.FieldTypeEnum",
      "name": "fieldType",
      "question": true,
      "stem": "fieldtyp",
    },
    Object {
      "kind": "string",
      "name": "indexAnalyzer",
      "question": true,
      "stem": "indexanalyz",
    },
    Object {
      "kind": "string",
      "name": "searchAnalyzer",
      "question": true,
      "stem": "searchanalyz",
    },
    Object {
      "kind": "Field.SimilarityEnum",
      "name": "similarity",
      "question": true,
      "stem": "similar",
    },
  ],
}
`;

exports[`#parseInterface should parse the interfaces correctly: example4 1`] = `
Object {
  "name": "Example4",
  "structure": Array [
    Object {
      "kind": "number",
      "name": "minZoom",
      "question": true,
      "stem": "minzoom",
    },
    Object {
      "kind": "number",
      "name": "maxZoom",
      "question": true,
      "stem": "maxzoom",
    },
    Object {
      "kind": "number",
      "name": "radius",
      "question": true,
      "stem": "radiu",
    },
    Object {
      "kind": "number",
      "name": "extent",
      "question": true,
      "stem": "extent",
    },
    Object {
      "kind": "number",
      "name": "nodeSize",
      "question": true,
      "stem": "nodes",
    },
    Object {
      "kind": "boolean",
      "name": "log",
      "question": true,
      "stem": "log",
    },
    Object {
      "kind": "(accumulated:any,props:any)=>void",
      "name": "reduce",
      "question": true,
      "stem": "reduc",
    },
    Object {
      "kind": "()=>any",
      "name": "initial",
      "question": true,
      "stem": "initi",
    },
    Object {
      "kind": "(props:any)=>any",
      "name": "map",
      "question": true,
      "stem": "map",
    },
  ],
}
`;
